File: src/python/challenges/problems/minimum_genetic_mutation_test.py

# https://leetcode.com/problems/minimum-genetic-mutation

from typing import List
from collections import deque
from core.problem_base import *

class MinimumGeneticMutation(ProblemBase):
    def Solution(self, startGene: str, endGene: str, bank: List[str]) -> int:
        seen = {}
        queue = deque([(endGene, len(endGene) - 1, 0)])

        result = -1
        while queue:
            gene, index, count = queue.popleft()

            if gene == startGene:
                result = count
                break

            if gene not in bank:
                continue

            if gene in seen:
                index -= 1
            else:
                count += 1
                index = len(endGene) - 1
                seen[gene] = count

            for l in ['A', 'C', 'G', 'T']:
                queue.append((gene[:index] + l + gene[index + 1:], index, seen[gene]))

        return result

if __name__ == '__main__':
    TestGen(MinimumGeneticMutation) \
        .Add(lambda tc: tc.Param("AAAAAAAA").Param("AAAAACGG").Param(["AAAAAAGA","AAAAAGGA","AAAAACGA","AAAAACGG","AAAAAAGG","AAAAAAGC"]).Result(3)) \
        .Add(lambda tc: tc.Param("AAAAAAAA").Param("CCCCCCCC").Param(["AAAAAAAA","AAAAAAAC","AAAAAACC","AAAAACCC","AAAACCCC","AACACCCC","ACCACCCC","ACCCCCCC","CCCCCCCA","CCCCCCCC"]).Result(8)) \
        .Add(lambda tc: tc.Param("AACCGGTT").Param("AACCGCTA").Param(["AACCGGTA","AACCGCTA","AAACGGTA"]).Result(2)) \
        .Add(lambda tc: tc.Param("AACCGGTT").Param("AAACGGTA").Param(["AACCGGTA","AACCGCTA","AAACGGTA"]).Result(2)) \
        .Add(lambda tc: tc.Param("AAAAAAAA").Param("CCCCCCCC").Param(["AAAAAAAA","AAAAAAAC","AAAAAACC","AAAAACCC","AAAACCCC","AACACCCC","ACCACCCC","ACCCCCCC","CCCCCCCA"]).Result(-1)) \
        .Add(lambda tc: tc.Param("AACCGGTT").Param("AACCGGTA").Param(["AACCGGTA"]).Result(1)) \
        .Run()

File: src/go/challenges/problems/minimum_genetic_mutation_test.go
// https://leetcode.com/problems/minimum-genetic-mutation

package problems

import (
	"container/list"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type MinimumGeneticMutation struct{}

func TestMinimumGeneticMutation(t *testing.T) {
	gen := core.TestSuite[MinimumGeneticMutation]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("AACCGGTT").Param("AACCGCTA").Param([]string{"AACCGGTA", "AACCGCTA", "AAACGGTA"}).Result(2)
	}).Run(t)
}

func (MinimumGeneticMutation) Solution(startGene string, endGene string, bank []string) int {
	bankSet := make(map[string]bool)
	for _, gene := range bank {
		bankSet[gene] = true
	}

	seen := make(map[string]int)
	queue := list.New()
	queue.PushBack(struct {
		gene  string
		index int
		count int
	}{endGene, len(endGene) - 1, 0})

	for queue.Len() > 0 {
		elem := queue.Front()
		queue.Remove(elem)
		current := elem.Value.(struct {
			gene  string
			index int
			count int
		})

		gene := current.gene
		index := current.index
		count := current.count

		if gene == startGene {
			return count
		}

		if !bankSet[gene] {
			continue
		}

		if _, exists := seen[gene]; exists {
			index--
		} else {
			count++
			index = len(endGene) - 1
			seen[gene] = count
		}

		if index < 0 {
			continue
		}

		for _, l := range []rune{'A', 'C', 'G', 'T'} {
			mutated := gene[:index] + string(l) + gene[index+1:]
			queue.PushBack(struct {
				gene  string
				index int
				count int
			}{mutated, index, seen[gene]})
		}
	}

	return -1
}
